### PR TITLE
Add threaded runner for the filter catalog

### PR DESCRIPTION
**Pull request by bp-kelley**
This adds a simple threaded runner to the filter catalog.

```
smiles = [...]
results = FilterCatalog.RunFilterCatalog(fc, smiles, numThreads=-1)
assert len(results) == len(smiles)
for result in results:
  if len(result): # filters have been hit
     filters = [x.GetDescription() for x in result]
```

The one thing I don't like is that we don't have a "bad smiles" filter, so smiles strings that can't be parsed return an empty filter list.

### CODE REVIEW

**Comment by bp-kelley** [comment]
The system returns a value for every molecule.  Mols with bad smiles strings or that don’t hit filters return empty vectors of matches.

My complaint is there should be a “bad smiles” filter here :)

----
Brian Kelley

**Comment by greglandrum** [comment]
> The system returns a value for every molecule. Mols with bad smiles strings or that don’t hit filters return empty vectors of matches. My complaint is there should be a “bad smiles” filter here :)

Ah, sure enough, I was reading it wrong. Sorry about that.
In that case there are just the few minor remarks remaining

File: Code/GraphMol/FilterCatalog/FilterCatalogRunner.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+#ifdef RDK_THREADSAFE_SSS  
+  if (numThreads == -1)
+    numThreads = (int)getNumThreadsToUse(numThreads);
+  else
+    numThreads = std::min(numThreads, (int)getNumThreadsToUse(numThreads));
+
+  std::vector<std::future<void>> thread_group;  
+  for (int thread_group_idx = 0; thread_group_idx < numThreads;
+       ++thread_group_idx) {
+    // need to use boost::ref otherwise things are passed by value
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/FilterCatalog/FilterCatalogRunner.cpp]: Grow isn't quite right, it just can't change memory location.

**Comment by bp-kelley** [comment]
Ok, I think this is ready.

**Comment by bp-kelley** [comment]
Extra super ready to go.

**Pull request merged**